sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -332,12 +332,14 @@ def trim_links(links, link_count_limit, link_attribute_count_limit) # rubocop:di
           return nil if links.nil?
 
           if links.size <= link_count_limit &&
-             links.all? { |link| link.attributes.size <= link_attribute_count_limit && Internal.valid_attributes?(name, 'link', link.attributes) }
+             links.all? { |link| link.span_context.valid? && link.attributes.size <= link_attribute_count_limit && Internal.valid_attributes?(name, 'link', link.attributes) }
             return links.frozen? ? links : links.clone.freeze
           end
 
           # Slow path: trim attributes for each Link.
-          links.last(link_count_limit).map! do |link|
+          valid_links = links.select { |link| link.span_context.valid? }
+          valid_links.shift(links.size - link_count_limit) if valid_links.size > link_count_limit
```

Comment:
There's no expectation of *which* links are dropped, if too many are provided. It is more efficient to drop the end of the array (because we don't need to move anything down).
```suggestion
          valid_links = links.select { |link| link.span_context.valid? }
          excess_link_count = valid_links.size - link_count_limit
          valid_links.pop(excess_link_count) if excess_link_count.positive?
```